File: experimentos/experimento/dp/suboperaciones.js
import { OPERACION, RESULTADO } from "./enums.js";
import { Estado } from "./estado.js";
import { CantidadExtraccionesHoy, DifFechasDias, ExisteUsuarioDNIConClave, HayCambioDeClaveEsteMes, LONGITUD, PerteneceACONTIENE_LETRA_NUM } from "./functions.js";
import * as global from "./globals.js";

export function ConsultaSaldoOK(dni, clave ) {
    if(ExisteUsuarioDNIConClave(dni, clave)) { 
        let estado = Estado.cargar();

        return {
            "saldo": estado.saldos[dni],
            "res": RESULTADO.ok
        }
    }

}

export function ExtraccionOK(dni, clave, monto) {
    let estado = Estado.cargar();

    if (ExisteUsuarioDNIConClave(dni, clave) &&
    CantidadExtraccionesHoy(dni) <= 2 &&
    +monto <= +estado.sueldos[dni] / 2 &&
    +monto <= +estado.saldos[dni]) {
        estado.saldos[dni] = estado.saldos[dni] - monto;
        
        if(!estado.movimientos[dni]) { 
            estado.movimientos[dni] = {};
        }
        
        estado.movimientos[dni][global.ahora.toJSON()] = OPERACION.extraccion;
        estado.guardar();

        return RESULTADO.ok;
    } 
}

export function CambioClaveOK(dni, clave, nuevaClave) {
    let estado = Estado.cargar();

    if (ExisteUsuarioDNIConClave(dni, clave) &&
    LONGITUD(nuevaClave) >= global.LONG_MIN_CLAVE &&
    PerteneceACONTIENE_LETRA_NUM(nuevaClave) &&
    HayCambioDeClaveEsteMes(dni)) {
        estado.claves[dni] = nuevaClave;
        estado.movimientos[dni][global.ahora.toJSON()] = OPERACION.clave;
        estado.guardar();

        return RESULTADO.ok;
    }
}

export function ConsultaMovimientosOK(dni, clave, dniConsulta, desde, hasta) {
    let estado = Estado.cargar();

    if(dni == global.administrador && 
        estado.claves[dni] == clave &&
        estado.usuarios[dniConsulta]) {
            return {
                "movimientos": Object.getOwnPropertyNames(estado.movimientos[dniConsulta])
                .filter(x => DifFechasDias(x, desde) >= 0 && DifFechasDias(hasta, x))
                .map(f => {
                    debugger;
                    return f +": "+ estado.movimientos[dniConsulta][f];
                }),
                "res": RESULTADO.ok
            }
        }
}

export function AltaUsuarioOK(dniAdministrador, claveAdministrador, dni, clave, nombre, sueldo) {
    let estado = Estado.cargar();

    if(dniAdministrador == global.administrador &&
        estado.claves[dniAdministrador] == claveAdministrador &&
        !estado.usuarios[dni] &&
        Object.getOwnPropertyNames(estado.usuarios).length < 5) {
            estado.movimientos[dni] = {};
            estado.usuarios[dni] = nombre;
            estado.claves[dni] = clave;
            estado.saldos[dni] = sueldo;
            estado.sueldos[dni] = sueldo;

            estado.guardar();
            
            return RESULTADO.ok;
        }
}

export function CargaOK(dniAdministrador, claveAdministrador, saldo) {
    let estado = Estado.cargar();

    if(dniAdministrador == global.administrador &&
        estado.claves[dniAdministrador] == claveAdministrador) {
            estado.saldo = +estado.saldo + +saldo;
            estado.guardar();
            
            return RESULTADO.ok;
        }
}

File: experimentos/experimento/gt/Suboperaciones.cs
using System;
using System.Collections.Generic;
using System.Linq;
using static Solucion.Estado;

namespace Solucion
{
    public static class Suboperaciones
    {
        public static (RESULTADO, int) ConsultaSaldo(Estado estado, int dni, string clave)
        {
            int saldo = 0;
            if (!estado.usuarios.Keys.Contains(dni)) return (RESULTADO.usuarioInexistente, saldo);
            if (!(estado.claves[dni] == clave)) return (RESULTADO.claveIncorrecta, saldo);

            saldo = estado.saldos[dni];

            return (RESULTADO.ok, saldo);
        }

        public static RESULTADO Extraccion(Estado estado, int dni, string clave, int monto)
        {
            if (!estado.usuarios.Keys.Contains(dni)) 
                return RESULTADO.usuarioInexistente;

            if (!(estado.claves[dni] == clave)) 
                return RESULTADO.claveIncorrecta;

            if (!(estado.movimientos[dni]
                .Where(o => FuncionesExternas.MISMO_DIA(o.Key, Globales.ahora))
                .Where(o => o.Value == OPERACION.extraccion)
                .Count() <= 2))
                return RESULTADO.noCumplePoliticaExtraccion;

            if (!(monto <= estado.sueldos[dni] / 2)) return RESULTADO.noCumplePoliticaExtraccion2;

            if (!(monto <= estado.saldo)) return RESULTADO.saldoCajeroInsuficiente;

            estado.saldo -= monto;

            estado.saldos[dni] -= monto;

            var ahora = Globales.ahora;

            if (!estado.movimientos[dni].Any(o => o.Key == ahora))
                estado.movimientos[dni].Add(ahora, OPERACION.extraccion);

            return RESULTADO.ok;
        }

        public static RESULTADO CambioClave(Estado estado, int dni, string clave, string nueva_clave)
        {
            if (!estado.usuarios.Keys.Contains(dni))
                return RESULTADO.usuarioInexistente;

            if (!(estado.claves[dni] == clave))
                return RESULTADO.claveIncorrecta;

            if (!(FuncionesExternas.LONGITUD(nueva_clave) >= Globales.LONG_MIN_CLAVE))
                return RESULTADO.noCumpleRequisitosClave1;

            if (!(FuncionesExternas.CONTIENE_LETRA_NUM(nueva_clave)))
                return RESULTADO.noCumpleRequisitosClave2;

            if (estado.movimientos[dni]
                .Any(o => FuncionesExternas.MISMO_MES(o.Key, Globales.ahora) && o.Value == OPERACION.clave))
                return RESULTADO.cambioDeClaveBloqueado;

            estado.claves[dni] = nueva_clave;

            var ahora = Globales.ahora;

            if (!estado.movimientos[dni].Any(o => o.Key == ahora))
                estado.movimientos[dni].Add(ahora, OPERACION.clave);

            return RESULTADO.ok;
        }

        public static (RESULTADO, List<(DateTime,OPERACION)>) ConsultaMovimientos(Estado estado, int dni, string clave, int dni_consulta, 
            DateTime desde, DateTime hasta)
        {
            if (!(Globales.administrador == dni))
                return (RESULTADO.usuarioNoHabilitado, null);

            if (!(estado.claves[dni] == clave))
                return (RESULTADO.claveIncorrecta, null);

            if (!estado.usuarios.Keys.Contains(dni_consulta))
                return (RESULTADO.usuarioInexistente, null);

            var movimientos = estado.movimientos[dni_consulta].Where(o => FuncionesExternas.DIF_FECHAS_DIAS(o.Key, desde) >= 0 &&
                FuncionesExternas.DIF_FECHAS_DIAS(hasta, o.Key) >= 0).Select(o => (o.Key, o.Value));

            return (RESULTADO.ok, movimientos.ToList());
        }

        public static RESULTADO AltaUsuario(Estado estado, int dni_administrador, string clave_administrador, int dni,
            string clave, string nombre, int sueldo)
        {
            if (!(Globales.administrador == dni_administrador))
                return RESULTADO.usuarioNoHabilitado;

            if (!(estado.claves[dni_administrador] == clave_administrador))
                return RESULTADO.claveIncorrecta;

            if (estado.usuarios.Keys.Contains(dni))
                return RESULTADO.usuarioYaExistente;

            if (!(estado.usuarios.Count < 5))
                return RESULTADO.limiteUsuariosAlcanzado;

            estado.movimientos.Add(dni, new Dictionary<DateTime, OPERACION>());

            estado.usuarios.Add(dni, nombre);

            estado.claves.Add(dni, clave);

            estado.saldos.Add(dni, sueldo);

            estado.sueldos.Add(dni, sueldo);

            return RESULTADO.ok;
        }

        public static RESULTADO Carga(Estado estado, int dni_administrador, string clave_administrador, int saldo)
        {
            if (!(Globales.administrador == dni_administrador))
                return RESULTADO.usuarioNoHabilitado;

            if (!(estado.claves[dni_administrador] == clave_administrador))
                return RESULTADO.claveIncorrecta;

            estado.saldo += saldo;

            return RESULTADO.ok;
        }
    }
}
